test/test_helpers.py

```diff
@@ -74,6 +77,95 @@ def fake_script_calls(test_case: unittest.TestCase,
     return calls  # type: ignore
 
 
+def create_framework(
+        monkeypatch: pytest.MonkeyPatch,
+        request: pytest.FixtureRequest,
+        *,
+        meta: typing.Union[ops.CharmMeta, None] = None):
+    monkeypatch.setenv("PATH", str(pathlib.Path(__file__).parent / 'bin'), prepend=os.pathsep)
+    monkeypatch.setenv("JUJU_UNIT_NAME", "local/0")
+
+    tmpdir = pathlib.Path(tempfile.mkdtemp())
+    request.addfinalizer(functools.partial(shutil.rmtree, str(tmpdir)))
+
+    class CustomEvent(ops.EventBase):
+        pass
+
+    class TestCharmEvents(ops.CharmEvents):
+        custom = ops.EventSource(CustomEvent)
+
+    # Relations events are defined dynamically and modify the class attributes.
+    # We use a subclass temporarily to prevent these side effects from leaking.
+    ops.CharmBase.on = TestCharmEvents()  # type: ignore
+
+    def cleanup():
+        ops.CharmBase.on = ops.CharmEvents()  # type: ignore
+
+    request.addfinalizer(cleanup)
+
+    if not meta:
+        meta = ops.CharmMeta()
+    model = ops.Model(meta, _ModelBackend('local/0'))  # type: ignore
+    # we can pass foo_event as event_name because we're not actually testing dispatch
+    framework = ops.Framework(SQLiteStorage(':memory:'), tmpdir, meta, model)  # type: ignore
+    request.addfinalizer(framework.close)
+
+    return framework
+
+
+class FakeScript:
+    def __init__(self,
+                 monkeypatch: pytest.MonkeyPatch,
+                 tmp_path: pathlib.Path,
+                 fake_script_path: typing.Union[pathlib.Path, None] = None):
+        self.monkeypatch = monkeypatch
+        if fake_script_path is None:
+            self.fake_script_path = tmp_path
+            self.monkeypatch.setenv(
+                "PATH",
+                str(self.fake_script_path),
+                prepend=os.pathsep)  # type: ignore
+        else:
+            self.fake_script_path = fake_script_path
+
+    def write(self,
+              name: str,
+              content: str):
+        template_args: typing.Dict[str, str] = {
+            'name': name,
+            'path': self.fake_script_path.as_posix(),  # type: ignore
+            'content': content,
+        }
+
+        path: pathlib.Path = self.fake_script_path / name  # type: ignore
+        with path.open('wt') as f:  # type: ignore
+            # Before executing the provided script, dump the provided arguments in calls.txt.
+            # ASCII 1E is RS 'record separator', and 1C is FS 'file separator', which
+            # seem appropriate.
+            f.write(  # type: ignore
+                '''#!/bin/sh
+    {{ printf {name}; printf "\\036%s" "$@"; printf "\\034"; }} >> {path}/calls.txt
```

Comment:
This is probably also in the original, but I think it's a bit confusing that we reference the characters by the hex value in the comment and then use the octal value in printf. Maybe just `RS 'record separator' (octal 036 in ASCII)` or something like that would be clearer to readers?